src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -298,19 +305,62 @@ public void PushFiles(string localDirectory, string devicePath, bool removeIfPre
             _log.LogDebug($"Copied {filesToCopy.Length} files");
         }
 
-        public (string StandardOutput, string StandardError, int ExitCode) RunApkInstrumentation(string apkName, TimeSpan timeout)
+        public Dictionary<string, string> GetAttachedDevicesAndArchitectures()
         {
-            return RunApkInstrumentation(apkName, "", new Dictionary<string, string>(), timeout);
+            Dictionary<string, string> devicesAndArchitectures = new Dictionary<string, string>();
+
+            var result = RunAdbCommand("devices -l");
+            string standardOutput = result.StandardOutput;
+
+            // If the adb server is dead, this can return 0 exit code and no output; retry a few times if so)
+            int retriesLeft = 5;
+            while (retriesLeft-- > 0 && result.ExitCode == (int)AdbExitCodes.SUCCESS && standardOutput == "")
+            {
+                result = RunAdbCommand("devices -l");
+                standardOutput = result.StandardOutput;
+            }
+
+            if (result.ExitCode == (int)AdbExitCodes.SUCCESS)
+            {
+                string[] lines = standardOutput.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries);
+                
+                for (int lineNumber = 1; lineNumber < lines.Length; lineNumber++)
+                {
+                    _log.LogDebug($"Evaluating line: {lines[lineNumber]}");
+                    var lineParts = lines[lineNumber].Split(' ');
+                    if (lineParts.Length > 0)
+                    {
+                        var shellArchitecture = RunAdbCommand($"-s {lineParts[0]} shell getprop ro.product.cpu.abi");
+                        var deviceSerial = lineParts[0];
+
+                        if (shellArchitecture.ExitCode == (int)AdbExitCodes.SUCCESS)
+                        {
+                            devicesAndArchitectures.Add(deviceSerial, shellArchitecture.StandardOutput.Trim());
+                        }
+                        else
+                        {
+                            _log.LogError($"Error trying to get device architecture: {shellArchitecture.StandardError}");
+                            devicesAndArchitectures.Add(deviceSerial, "unknown");

```

Comment:
It might be better to have a `Dictionary<string, string?>` and put null where it's unknown and the nullability will later force you to handle that case somewhere in the code rather than having to remember to handle the "unknown" string.